src/Utils/proto-utils.ts

```diff
@@ -0,0 +1,23 @@
+import { proto } from '../../WAProto'
+import { BufferJSON } from './generics'
+
+type ProtoMessageClass<T = any> = {
+	new (...args: any[]): T
+	decode(buffer: Uint8Array | Buffer): T
+}
+
+type DecodableProtoKey = {
+	[K in keyof typeof proto]: (typeof proto)[K] extends ProtoMessageClass ? K : never
+}[keyof typeof proto]
+
+export function decodeAndHydrate<T extends DecodableProtoKey>(
+	type: T,
+	buffer: Uint8Array | Buffer
+): InstanceType<(typeof proto)[T]> {
+	const MessageType = proto[type]
```

Comment:
why do this when you can simply do 

`decodeAndHydrate(proto.Message, buffer)`

Reply:
It allows us to autocomplete and reference while developing, and it also allows for more accurate type definition when released. I'm sure you can make it work